viewer/packages/rendering/src/render-pipelines/CadGeometryRenderPipeline.ts

```diff
@@ -0,0 +1,125 @@
+/*!
+ * Copyright 2022 Cognite AS
+ */
+
+import * as THREE from 'three';
+import { CadMaterialManager } from '../CadMaterialManager';
+import { GeometryPass } from '../render-passes/GeometryPass';
+import { RenderMode } from '../rendering/RenderMode';
+import { RenderOptions } from '../rendering/types';
+import { RenderPass } from '../RenderPass';
+import { RenderPipelineProvider } from '../RenderPipelineProvider';
+import { createRenderTarget, setupCadModelsGeometryLayers } from '../utilities/renderUtilities';
+import { IdentifiedModel } from '../utilities/types';
+import { CadGeometryRenderTargets } from './types';
+
+type CadGeometryRenderPasses = {
+  back: GeometryPass;
+  ghost: GeometryPass;
+  inFront: GeometryPass;
+};
+
+type RenderState = {
+  autoClear: boolean;
+  clearColor: THREE.Color;
+  clearAlpha: number;
+};
+
+export class CadGeometryRenderPipeline implements RenderPipelineProvider {
+  private readonly _cadGeometryRenderTargets: CadGeometryRenderTargets;
+  private readonly _cadGeometryRenderPasses: CadGeometryRenderPasses;
+  private readonly _cadModels: IdentifiedModel[];
+  private _currentRendererState: RenderState;
+  private readonly _materialManager: CadMaterialManager;
+
+  get cadGeometryRenderTargets(): CadGeometryRenderTargets {
+    return this._cadGeometryRenderTargets;
+  }
+
+  constructor(
+    scene: THREE.Scene,
+    cadModels: IdentifiedModel[],
+    materialManager: CadMaterialManager,
+    renderOptions: RenderOptions
+  ) {
+    this._cadModels = cadModels;
+    this._materialManager = materialManager;
+
+    const multisampleCount = renderOptions.multiSampleCountHint ?? 0;
+    this._cadGeometryRenderTargets = this.initializeRenderTargets(multisampleCount);
+    this._cadGeometryRenderPasses = this.initializeRenderPasses(scene);
+  }
+
+  public *pipeline(renderer: THREE.WebGLRenderer): Generator<RenderPass> {
+    this.pipelineSetup(renderer);
+
+    renderer.setRenderTarget(this._cadGeometryRenderTargets.back);
+    yield this._cadGeometryRenderPasses.back;
+
+    renderer.setRenderTarget(this._cadGeometryRenderTargets.ghost);
+    yield this._cadGeometryRenderPasses.ghost;
+
+    renderer.setRenderTarget(this._cadGeometryRenderTargets.inFront);
+    yield this._cadGeometryRenderPasses.inFront;
+
+    this.pipelineTearDown(renderer);
+  }
+
+  public dispose(): void {}
+
+  private pipelineSetup(renderer: THREE.WebGLRenderer) {
+    this._currentRendererState = {
+      autoClear: renderer.autoClear,
+      clearColor: renderer.getClearColor(new THREE.Color()),
+      clearAlpha: renderer.getClearAlpha()
+    };
```

Comment:
Are you aware of `WebGLRendererStateHelper` that helps to store/restore state for WebGLRenerer?